test/rules-ceddl-fullstory.spec.ts

```diff
@@ -1,149 +1,159 @@
-    expectMatch(payload, basicDigitalData.product[0].category, 'primaryCategory');
-    expectEqual(payload.customProp, 'Foo');
-  });
-
-  it('it should send CEDDL cart to FS.event', () => {
-    expectGlobal('digitalData').cart.promotion = 'LaborDay2020'; // inject custom property
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-cart')], readOnLoad: true,
-    });
-
-    const [eventName, payload] = expectFS('event');
-    expectEqual(eventName, 'cart');
-    expectEqual(payload.cartID, basicDigitalData.cart.cartID);
-    expectEqual(payload.basePrice, basicDigitalData.cart.price.basePrice);
-    // @ts-ignore custom property
-    expectEqual(payload.promotion, 'LaborDay2020');
-  });
-
-  it('it should send dynamic CEDDL cart item additions to FS.event', () => {
-    const firstProduct = basicDigitalData.product[0];
-
-    const secondProduct = {
-      ...firstProduct,
-      productInfo: {
-        ...firstProduct.productInfo,
-        sku: 'test',
-      },
-    };
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-cart-item')],
-    });
-
-    expectGlobal('digitalData').cart.item.push(firstProduct);
-
-    let [eventName, payload] = expectFS('event');
-    expectEqual(eventName, 'cart_item');
-    expectEqual(payload.productInfo.sku, firstProduct.productInfo.sku);
-    expectUndefined(payload, 'linkedProduct');
-
-    expectGlobal('digitalData').cart.item.push(secondProduct);
-
-    [eventName, payload] = expectFS('event');
-    expectEqual(eventName, 'cart_item');
-    expectEqual(payload.productInfo.sku, secondProduct.productInfo.sku);
-    expectUndefined(payload, 'linkedProduct');
-  });
-
-  it('it should not send CEDDL cart item products to FS.event on load', () => {
-    expectGlobal('digitalData').cart.item.push(basicDigitalData.product[0]);
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-cart-item')],
-    });
-
-    const fs = expectGlobal('FS');
-    expectNoCalls(fs, 'event');
-  });
-
-  it('it should send CEDDL page properties to FS.event', () => {
-    expectGlobal('digitalData').page.framework = 'react'; // inject custom property
-
-    ExpectObserver.getInstance().create({
-      rules: [expectRule('fs-event-ceddl-page')], readOnLoad: true,
+  getRulesetTestEnvironments().forEach((testEnv) => {
+    describe(`test environment: ${testEnv.name}`, () => {
+      let testHarness: RulesetTestHarness;
+
+      beforeEach(async () => {
+        testHarness = await testEnv.createTestHarness(ceddlRules, { digitalData: basicDigitalData });
+      });
+
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
+
+      after(async () => {
+        await testEnv.tearDown();
+      });
+
+      it('sends the first CEDDL product to FS.event', async () => {
+        await testHarness.execute(() => {
+          (globalThis as any).digitalData.product[0].attributes = { customProp: 'foo' };
+        });
```

Comment:
Previous CEDDL ruleset tests specified `readOnLoad: true` which was testing a different code path (direct operator handler invocation) instead of the code path used by the CEDDL data layer integration (handler triggering via event listeners). These tests now all test dynamic data layer changes via event listeners which exercises the code paths used by the CEDDL data layer integration.